### PR TITLE
Escape "query" parameter in JSON configuration

### PR DESCRIPTION
**Pull request by JanPetr**
Related Issue(s): #769

### CODE REVIEW

File: view/frontend/web/internals/common.js [hunk 1]
```diff
@@ -834,3 +834,20 @@ var AlgoliaBase64 = {
...
 	}
 };
+
+var htmlspecialchars_decode = function(string) {
+	var unescapedString = string,
+		specialchars = [
+			[ '"', '&quot;' ],
+			[ '>', '&gt;' ],
+			[ '<', '&lt;' ],
+			[ '&', '&amp;' ]
```
Comment (tkrugg) [inline view/frontend/web/internals/common.js]: it is usually expected to include `'`. Look for example in this lodash implementation: https://github.com/lodash/lodash/blob/master/unescape.js#L3-L7

File: Block/Configuration.php [hunk 2]
```diff
@@ -186,7 +186,7 @@ public function getConfiguration()
             'maxValuesPerFacet' => (int) $config->getMaxValuesPerFacet(),
             'autofocus' => true,
             'request' => [
-                'query' => html_entity_decode($query),
+                'query' => htmlspecialchars(html_entity_decode($query)),
```
Comment (tkrugg) [inline Block/Configuration.php]: shouldn't the function here be `htmlspecialchars_decode ` ?

Comment (bsuravech) [reply Block/Configuration.php]: `htmlspecialchars` is correct here as we should be encoding it at this stage and decoding in js.

**Comment by bsuravech** [comment]
@tkrugg I don't know if we want to include lodash here just for this fix. I've made some adjustments so that the method is defined with in the algolia object. Can you review that when you have a moment? I've also added the apostrophe character "`'`"

Can you elaborate on your comment regarding "`htmlspecialchars` looks not defined." ? That is a php function that encodes the query in the configuration json output.

File: view/frontend/web/internals/common.js [hunk 3]
```diff
@@ -45,6 +45,23 @@ var algolia = {
...
+	},
+	htmlspecialcharsDecode: function(string) {
+		var unescapedString = string,
+			specialchars = [
+				[ '"', '&quot;' ],
+				[ '>', '&gt;' ],
+				[ '<', '&lt;' ],
+				[ '&', '&amp;' ],
+				[ "'", '&#39;' ]
+			];
```
Comment (broujo) [inline view/frontend/web/internals/common.js]: Hey team,

The OWASP recommends to escape forward slash too (https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html). Not sure how necessary it is to actively prevent XSS, but I think it's still worth including.

File: Block/Configuration.php (hunk 2 above)
Comment (tkrugg) [reply Block/Configuration.php]: yes, it makes sense.

**Comment by tkrugg** [comment]
@bsuravech sorry for the late reply.
> htmlspecialchars looks not defined.
I just meant I didn't find the definition of the function. Can you point it to me?

Regarding lodash, as far as I can tell, it is already part of the scope so you might as well just use it, and their function are always battle tested, but I don't have a strong opinion on this.

**Comment by bsuravech** [comment]
Closing out as it is outdated. If issue returns, we'll address it in the latest versions.

**Pull request closed without merging**